File: src/main/java/stni/atlassian/remote/queryold/DefaultQueryService.java
package stni.atlassian.remote.queryold;

import com.atlassian.jira.rpc.soap.beans.RemoteIssue;
import com.atlassian.jira.rpc.soap.beans.RemoteProject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import stni.atlassian.remote.JiraService;

import java.util.Collections;
import java.util.List;

/**
 *
 */
public class DefaultQueryService implements QueryService {
    private final static Logger LOG = LoggerFactory.getLogger(DefaultQueryService.class);

    private final JiraService service;

    public DefaultQueryService(JiraService service) {
        this.service = service;
    }

    public List<RemoteProject> getProjectsByKey(String... keys) {
        if (keys.length == 0) {
            return service.getAllProjects();
        }
        if (keys.length == 1) {
            return Collections.singletonList(service.getProjectByKey(keys[0]));
        }
        return service.getProjectsByKey(keys);
    }

    public RemoteIssue getIssue(String issueKey) {
        return service.getIssue(issueKey);
    }

    @Override
    public RemoteIssue[] getIssuesFromFilter(String filter) {
        return service.getIssuesFromFilter(filter);
    }

    public RemoteIssue[] getIssuesFromJqlSearch(String query, int maxResults) {
        LOG.debug("Executing jql '{}'", query);
        return service.getIssuesFromJqlSearch(query, maxResults);
    }

    public String getBaseUrl() {
        return service.getBaseUrl();
    }

    public String getCustomField(RemoteIssue issue, String name) {
        return service.customFieldByName(issue, name);
    }

}

File: src/main/scala/stni/atlassian/remote/query/DefaultQueryService.scala
package stni.atlassian.remote.query

import org.slf4j.LoggerFactory
import com.atlassian.jira.rpc.soap.beans.{RemoteIssueType, RemotePriority, RemoteIssue, RemoteProject}
import java.util.Collections
import collection.convert.Wrappers
import stni.atlassian.remote.jira.JiraService
import stni.atlassian.remote.AtlassianException

/**
 *
 */
class DefaultQueryService(service: JiraService) extends QueryService {
  private val log = LoggerFactory.getLogger(getClass)

  def getProjectsByKey(keys: String*): Seq[RemoteProject] = {
    Wrappers.JListWrapper(
      keys.length match {
        case 0 => wrap((dummy: String) => service.getAllProjects, null)
        case 1 => Collections.singletonList(wrap(service.getProjectByKey, keys(0)))
        case _ => wrap((keys:Seq[String])=>service.getProjectsByKey(keys:_*), keys)
      })
  }

  def getIssue(issueKey: String): RemoteIssue = wrap(service.getIssue, issueKey)

  def getIssuesFromFilter(filter: String): Seq[RemoteIssue] = wrap(service.getIssuesFromFilter, filter)

  def getIssuesFromJqlSearch(query: String, maxResults: Int): Seq[RemoteIssue] = {
    log.debug("Executing jql '{}'", query)
    wrap((q: String) => service.getIssuesFromJqlSearch(q, maxResults), query)
  }

  def baseUrl: String = service.getBaseUrl

  def customField(issue: RemoteIssue, name: String): String = wrap((i: RemoteIssue) => service.customFieldByName(i, name), issue)

  def priorityById(id: String): RemotePriority = wrap(service.priorityById, id)

  def issueTypeById(id: String): RemoteIssueType = wrap(service.issueTypeById, id)

  def timeToResolve(issue: RemoteIssue): Long = wrap(service.timeToResolve, issue)

  private def wrap[P, T](exec: P => T, param: P): T =
    try {
      exec(param)
    } catch {
      case e: AtlassianException => throw new QueryException(param.toString, e.getCause)
    }
}
